### PR TITLE
Development

### PR DESCRIPTION
**Pull request by Jood80**
#### in this sprint:
- [x] add new controller:
   - add new user "via admin":  `POST`- `/add-user`
   - block user:  `PATCH` -  `/block-user/:userId`
   - edit user's profile: `PATCH` -  `/edit-profile`
   - get all content "paginated":  `GET` - `/contents`
   - get all users "paginated":  `GET` - `/users`
   - match user with his/her content: `PATCH` - `/match-user-content`
    - login via google: `POST` -  `/log/google` & `POST` - `/sign/google`
- [x] move all queries onto `services` folder    
- [x] use `morgan` & `winston` for logging. 
- [x] upload edited profile picture to s3
- [x] update email service with create user template and remove verify-email step
- [x] change all statics messages & values to comes from `constants` file
- [x] add extra validation layer(helper) over received schema requests.   
- [x] move all validation checks to helpers folder:
   -  checkUserExistance 
   -   comming soon ...
- [x] add a middleware to check current user role.
- [x] add `umzug`  to generate migrations and remove the old code used to generate dummy data
- [x] embed our `db-models-nc` npm package that we cloned from our neighbor [db-models-nc repo](https://github.com/NextUp-comedy/db-models-nc)
- update `package.json` scripts with: 
  - `build:db`: Joker script; use it to drop your db if exists, then build your database & seeding it with fake data. 
  - `migrate:up`: 
  - `migrate:down`
  - `migrate:reset`:
 ------- 
 ### How to test it over:
 1. To run the project"
 - `$ npm i`
 - in your `.env.development ` add your `env vars`
 - `$ npm run build:db`
 - `$ npm run dev`
------
2. after successfully running your code, you should have this dummy data generated in your db

**Whereas** :

```javascript
  const user_status_id = {
        PENDING: 1, 
        APPROVED: 2,
        REJECTED: 3,
        BANNED: 4
    }
 ``` 
``` 
admin - admin@gmail.com   | admin - approved 
zaher -  zaherabuamro@gmail.com  |  "comedian - approved"
ibrahim - ibrahim.jarada@gmail.com | "comedian - rejected"
ayman - ayman@gmail.com   | "comedian - banned"
rand - rand@gmail.com     |  "comedian - pending"
```
------
3. to interact with the process front & back:
    1. go to [`front-nc repo`](https://github.com/NextUp-comedy/front-nc) `development` branch. "pull the latest version"
    2. check the current flow for now in the demo section:
------------------    
## Demo    
  https://www.loom.com/share/a8513b10cf324908bdcdc22f394011b4
  https://www.loom.com/share/28f1305941b340e4afde518c239319c6

### CODE REVIEW

**Review by NoureldeanSaed — APPROVED** [review]

**Review by mohanagy — APPROVED** [review]

File: src/controllers/auth/forgetPassword.ts [hunk 1]
```diff
@@ -1,42 +1,47 @@
...
-import { constants, checkExistence } from '../../helpers';
+import {
+  constants, checkExistence, signToken, sendEmail,
+} from '../../helpers';
+import { getUserByEmail } from '../../services';
 
 export default async ({ body }: Request, res: Response, next: NextFunction)
 :Promise<void> => {
   const { resetToken } = constants.messages.token;
   const { emailCheck } = constants.messages.check;
```
Comment (mohanagy) [inline src/controllers/auth/forgetPassword.ts:11]: you can merge these two lines in one line, and destruct token and check.
then you can call resetToken by token and emailCheck by check
`  const { token,check } = constants.messages;`

File: src/controllers/auth/forgetPassword.ts [hunk 2]
```diff
@@ -1,42 +1,47 @@
...
+  const { OK } = constants.HttpStatus;
+  const { email } = body;
```
Comment (mohanagy) [inline src/controllers/auth/forgetPassword.ts:13]: Use DTO to extract whatever you want.
inside the DTO you can convert the email to lowercase.

File: src/controllers/auth/login.ts [hunk 3]
```diff
@@ -1,37 +1,44 @@
...
 Promise<void> => {
   const { email, password, rememberMe } = body;
   const { wrongEmailOrPassword, logIn } = constants.messages.authResponse;
   const { accessToken } = constants.messages.token;
+  const { OK, UNAUTHORIZED } = constants.HttpStatus;
   try {
-    await loginSchema.validateAsync(body);
     let expiresIn;
-    if (rememberMe) { expiresIn = '30d'; } else { expiresIn = '1h'; }
+    if (rememberMe) { expiresIn = '30d'; } else { expiresIn = '24h'; }
```
Comment (mohanagy) [inline src/controllers/auth/login.ts:16]: this is not the meaning of remember me!
search about it

File: src/controllers/auth/forgetPassword.ts [hunk 4]
```diff
@@ -1,42 +1,47 @@
...
+
+    await sendEmail({
+      email: lowerCaseEmail,
+      type: 'reset',
+      name,
+      redirectURL,
+    });
+
     res
       .cookie(resetToken, token)
```
Comment (mohanagy) [inline src/controllers/auth/forgetPassword.ts:41]: Why do we need to pass the token  in the cookies here ?

File: src/controllers/auth/signUpGoogle.ts [hunk 5]
```diff
@@ -1,59 +1,62 @@
...
-import { checkExistence } from '../../helpers';
+  checkExistence,
+  constants,
+  sendEmail,
+  googleAuthentication,
+} from '../../helpers';
+import { addUser } from '../../services';
 
-export default async ({ body }: Request, res: Response, next: NextFunction)
+export default async ({ body: { tokenId } }: Request, res: Response, next: NextFunction)
```
Comment (mohanagy) [inline src/controllers/auth/signUpGoogle.ts:12]: don't do this

File: src/controllers/auth/signUpGoogle.ts [hunk 6]
```diff
@@ -1,59 +1,62 @@
...
-      `${config.server.googleAPI}tokeninfo?id_token=${tokenId}`,
-    );
+      email, name, image, googleId,
+    } = await googleAuthentication(tokenId);
 
     await checkExistence.RegistrationCheck(email);
 
     const password = generatePassword.generate({
-      length: 20, numbers: true, strict: true, lowercase: true, uppercase: true,
+      length: 20,
```
Comment (mohanagy) [inline src/controllers/auth/signUpGoogle.ts:28]: Is the length 20 ? or 15?
in another place the password is 15

File: src/controllers/auth/signUpGoogle.ts [hunk 7]
```diff
@@ -1,59 +1,62 @@
...
-      googleId: sub,
+      googleId,
+      freeToBePaidRevenue: REVENUE_DEFAULT_VALUE,
+      accPaidRevenue: REVENUE_DEFAULT_VALUE,
     });
-    res.status(201).redirect(`${config.server.clientURL}/verifyEmail`);
+
+    await sendEmail({
+      email: user.email,
+      type: 'verify',
```
Comment (mohanagy) [inline src/controllers/auth/signUpGoogle.ts:52]: I think we can define the types as constants inside the constants file

File: src/controllers/auth/signup.ts [hunk 8]
```diff
@@ -1,52 +1,43 @@
...
 
     const hashedPassword = await hash(password, 10);
-    const user = await Users.create({
-      username,
-      email,
-      roleId: 2,
+
+    const user = await addUser({
+      name,
+      email: email.toLowerCase(),
```
Comment (mohanagy) [inline src/controllers/auth/signup.ts:22]: you already defined it `const lowercaseEmail = email.toLowerCase();`

File: src/controllers/users/getPaginatedContents.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { Request, Response, NextFunction } from 'express';
+import { getContents } from '../../services';
```
Comment (mohanagy) [inline src/controllers/users/getPaginatedContents.ts:2]: getContents related to what exactly?
the name is not clear

File: src/controllers/users/rejectUser.ts [hunk 10]
```diff
@@ -1,27 +1,32 @@
...
-} from '../../utilities';
-import { checkExistence, constants } from '../../helpers';
+import config from '../../config';
+import { checkExistence, constants, sendEmail } from '../../helpers';
 
 export default async (req: Request, res: Response, next: NextFunction)
 :Promise<void> => {
   const { userId } = req.params;
+  const redirectURL = `${config.server.CLIENT_URL}/faq`; // TODO: to be modified with FAQ page
+  const contactUs = `mailto:${config.email.NEXTUP_COMEDY_SUPPORT_EMAIL}`;
```
Comment (mohanagy) [inline src/controllers/users/rejectUser.ts:10]: I see there is no input comes from the controller to the definition , in this case , move these lines to the constants file

File: src/helpers/checkUserExistence.ts [hunk 11]
```diff
@@ -1,125 +1,98 @@
...
-    if (userExists.status === rejected) throw new CustomError(rejectedUser, 401);
+    if (userStatus === APPROVED) throw new CustomError(ALREADY_APPROVED, CONFLICT);
 
-    if (userExists.isVerified) throw new CustomError(verifiedUser, 401);
+    if (userStatus === REJECTED) throw new CustomError(ALREADY_REJECTED, UNAUTHORIZED);
 
-    if (userExists) throw new CustomError(waitApprove, 401);
+    if (userStatus === PENDING) throw new CustomError(messages.authResponse.PENDING, UNAUTHORIZED);
+
+    if (userStatus === BANNED) throw new CustomError(messages.authResponse.BANNED, UNAUTHORIZED);
```
Comment (mohanagy) [inline src/helpers/checkUserExistence.ts:34]: I can't accept these lines , there is a better way to handle it

File: src/helpers/checkUserExistence.ts [hunk 12]
```diff
@@ -1,125 +1,98 @@
...
+    const userStatus = userExists.userStatusId;
 
-    if (!userExists?.isVerified) throw new CustomError(verifiedUser, 401);
+    if (userStatus === REJECTED) throw new CustomError(ALREADY_REJECTED, UNAUTHORIZED);
 
-    if (userExists.status === rejected) throw new CustomError(rejectedUser, 401);
+    if (userStatus === PENDING) throw new CustomError(messages.authResponse.PENDING, UNAUTHORIZED);
 
-    if (userExists.status !== approved) throw new CustomError(approvedUser, 401);
+    if (userStatus === BANNED) throw new CustomError(messages.authResponse.BANNED, UNAUTHORIZED);
```
Comment (mohanagy) [inline src/helpers/checkUserExistence.ts:63]: same , there is a better way to handle it

**Review by Ibrahim-Jarada — APPROVED** [review]

**Pull request merged**